README.md
# Magic Mirror Module: Remote Control w/ RESTful API

This module for the [Magic Mirror²](https://github.com/DiscoverITLLC/MagicMirror) allows you to quickly shutdown your mirror through a web browser.
The website should work fine on any device (desktop, smart phone, tablet, ...).
Since we all want our [SD cards to live a long and prosper life](http://raspberrypi.stackexchange.com/a/383) we properly shut down before pulling the power plug everytime, am I right?
Additionally you can hide and show modules on your mirror and do other cool stuff.

![The Main Menu](.github/main.png)
![The Power Menu](.github/power.png)
![Hide and Show a Module](.github/hide_show_module.gif)

**New in Version 2.0.0:** The module now includes a more RESTful API for controlling all aspects of your mirror from other network-enabled devices and controllers--anything that can open a URL. See the [API README](API/README.md) for more info!

## Installation

### Quick install

If you followed the default installation instructions for the [Magic Mirror²](https://github.com/DiscoverITLLC/MagicMirror) project, you should be able to use the automatic installer.
The following command will download the installer and execute it:
```bash
bash -c "$(curl -s https://raw.githubusercontent.com/DiscoverITLLC/MMM-Remote-Control/master/installer.sh)"
```

### Manual install

- (1) Clone this repository in your `modules` folder, and install dependencies:
```bash
cd ~/MagicMirror/modules # adapt directory if you are using a different one
git clone https://github.com/DiscoverITLLC/MMM-Remote-Control.git
cd MMM-Remote-Control
npm install
```

- (2) Add the module to your `config/config.js` file, if you add a `position`, it will display the URL to the remote on the mirror.
```js
{
    module: 'MMM-Remote-Control',
    // uncomment the following line to show the URL of the remote control on the mirror
    // position: 'bottom_left',
    // you can hide this module afterwards from the remote control itself
    config: {
        customCommand: {},  // Optional, See "Using Custom Commands" below
        showModuleApiMenu: true, // Optional, Enable the Module Controls menu
        secureEndpoints: true, // Optional, See API/README.md
        // uncomment any of the lines below if you're gonna use it
        // customMenu: "custom_menu.json", // Optional, See "Custom Menu Items" below
        // apiKey: "", // Optional, See API/README.md for details
        // classes: {} // Optional, See "Custom Classes" below
    }
},
```

- (3) For security reasons, the MagicMirror (and therefore the Remote Control) is *not* reachable externally.
To change this, configure `address`, and `ipWhitelist` in your `config.js` (see [these lines in the sample config](https://github.com/DiscoverITLLC/MagicMirror/blob/master/config/config.js.sample#L12-L22)).
For example change `address` to `0.0.0.0` and add two allowed devices with IP-Adresses `192.168.0.42` and `192.168.0.50`:
```
    address : '0.0.0.0',
    port: 8080,
    ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.0.42", "::ffff:192.168.0.50"],"
```
You can also add multiple devices in an IP range (e.g. all devices with `192.168.0.X`):
```
    ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.0.1/120", "192.168.0.1/24"],
```

- (4) Restart your Magic Mirror² (i.e. `pm2 restart MagicMirror`).

- (5) Access the remote interface on [http://192.168.xxx.xxx:8080/remote.html](http://192.168.xxx.xxx:8080/remote.html) (replace with IP address of your RaspberryPi).

Note: If your user does not have `sudo` rights, the shutdown does not work (it *should* work for everyone who did not change anything on this matter).

### Update

Update this module by navigating into its folder on the command line and using `git pull`:

```bash
cd ~/MagicMirror/modules/MMM-Remote-Control # adapt directory if you are using a different one
git pull
npm install # install (new) dependencies
```

Alternatively you can run the `installer.sh` script again:
```bash
~/MagicMirror/modules/MMM-Remote-Control/installer.sh
```

### Select version manually

You can check out specific versions in the following way.
First look at which versions are available:
```bash
cd MagicMirror/modules/MMM-Remote-Control # or wherever you installed the Mirror and the module
git fetch # fetch all tags
git tag # display them
```

The output should look similar to this:
```
v1.0.0
v1.1.0
v1.1.1
v1.1.2
```
Then you can checkout that version with, for example `git checkout v1.0.0`, or use `git checkout master` to checkout the most recent version.

## Known limitations

Whenever you change the order of modules in `config.js` or add/remove modules, the indices of the modules change.
Therefore the hidden/shown status of modules might not be correctly applied.
If this happens, simply reconfigure and save it again.

## Call methods from other modules

You can call any of the methods provided in the UI directly through a GET request, or a module notification.
For example you can use [MMM-ModuleScheduler](https://forum.magicmirror.builders/topic/691/mmm-modulescheduler) to automatically shutdown your RasberryPi at a certain time, or integrate it with home automation systems. Or use  [MMM-Navigate](https://github.com/Ax-LED/MMM-Navigate) to allow direct actions from your Mirror by using a rotating button. 

### Examples

- Example for a REST API GET request to trigger a RaspberryPi restart:
```
http://192.168.xxx.xxx:8080/api/restart
```

- Example to trigger a RaspberryPi restart in your module:
```
this.sendNotification('REMOTE_ACTION', {action: 'RESTART'});
```

See some specific examples for controlling your mirror from other modules and add your own examples [in the Wiki page here](https://github.com/shbatm/MMM-Remote-Control/wiki/Examples-for-Controlling-from-Another-Module)

### List of actions

#### System Control:

| Action | Description |
| :-: | ------------- |
| SHUTDOWN | Shutdown your RaspberryPi |
| REBOOT | Restart your RaspberryPi |
| MONITORON | Switch your display on. Also sends a `"USER_PRESENCE": true` notification. |
| MONITOROFF | Switch your display off. Also sends a `"USER_PRESENCE": false` notification. |
| MONITORTOGGLE | Toggle the display on or off (with respective `"USER_PRESENCE"` notification. |
| MONITORSTATUS | Report back the monitor status (on or off) |

#### MagicMirror Control:

| Action | Description |
| :-: | ------------- |
| RESTART | Restart your MagicMirror |
| REFRESH | Refresh mirror page |
| UPDATE | Update MagicMirror and any of it's modules |
| SAVE | Save the current configuration (show and hide status of modules, and brightness), will be applied after the mirror starts |
| BRIGHTNESS | Change mirror brightness, with the new value specified by `value`. `100` equals the default, possible range is between `10` and `200`. |

#### MagicMirror Electron Browser Window Control:

| Action | Description |
| :-: | ------------- |
| MINIMIZE | Minimize the browser window. |
| TOGGLEFULLSCREEN | Toggle fullscreen mode on and off. |
| DEVTOOLS | Open the DevTools console window. |

#### Module Control:

| Action | Description |
| :-: | ------------- |
| HIDE | Hide a module, with the name (or identifier--see `MODULE_DATA` action) specified by `module` in the payload. You can also send `module: "all"` to hide all modules. |
| SHOW | Show a module (see above for how to specify which one). |
| TOGGLE | Toggle a module's visiblity (see above for how to specify which one). |
| FORCE | Force a module to show (see above for how to specify which one). |
| MODULE_DATA | Returns a JSON format of the data displayed in the UI, including all valid identifiers for the `HIDE` and `SHOW` action. |

#### Alerts and Notifications:

| Action | Description |
| :-: | ------------- |
| SHOW_ALERT | Show Default Alert/Notification |
| HIDE_ALERT | Hide Default Alert/Notification |
| USER_PRESENCE | Will send a notification "USER_PRESENCE" = true or false (according to "value" to all other modules. See examples above|
| NOTIFICATION | To send a notification to all modules, see the example in the [API README](API/README.md) |
| DELAYED | Send any of the above nested inside a "DELAYED" call to delay the action. Default is 10s. See Delayed Actions below. |

### Format of module data response

The response will be in the JSON format, here is an example:

```json
{
"moduleData":[
    {"hidden":false,"name":"alert","identifier":"module_0_alert"},
    {"hidden":true,"name":"clock","identifier":"module_1_clock","position":"bottom_right"},
    {"hidden":false,"name":"currentweather","identifier":"module_2_currentweather","position":"top_right"}
],
"brightness":40,
"settingsVersion":1
}
```

### Delayed Actions

You can delay an action by sending the notification nested inside of a `"DELAYED"` notification. Below is an example of turning off the monitor in 60s.

```js
this.sendSocketNotification("REMOTE_ACTION",
    {
        action: "DELAYED",
        did: "SOME_UNIQUE_ID", // Optional; Some string, in case you want to cancel later.
        timeout: 60,  // Optional; Default 10s
        abort: false, // Optional; send true to cancel an existing timer
        query: {
            action: "MONITOROFF"
        }
    });
```

Can also be used with the [API](https://documenter.getpostman.com/view/6167403/Rzfni66c) by adding `/delay?timeout=10s&did=something` to some routes.

### Using Custom Commands

Depending on your installation, some `shell` commands used by this module are not appropriate and can be overwritten by something that will work for you. To overwrite the commands, add a `customCommand` object to your config section.  The following commands are supported:

```js
    customCommand: {
        shutdownCommand: 'shell command to shutdown your pc',
        rebootCommand: 'shell command to reboot your pc',
        monitorOnCommand: 'shell command to turn on your monitor',
        monitorOffCommand: 'shell command to turn off your monitor',
        monitorStatusCommand: 'shell command to return status of monitor, must return either "HDMI" or "true" if screen is on; or "TV is Off" or "false" if it is off to be recognized'
    }
```

### Custom Classes

You probably wanna hide or show some modules at the same time, right? It's everything that we want this module for, of course.
Well, now you can add as many classes as you like, and define whether they show themself, hide or toggle between the two stages!

```js
    classes: {
        "Any Name You Want": {
            hide: ["calendar"],
            show: ["newsfeed"],
            toggle: ["clock"],
        },
        "Another Name You Want": {
            hide: ["newsfeed"],
            show: ["calendar"],
        },
    }
```

### Custom Menu Items

You can create your own customized menu items by providing creating a JSON file for the menu and providing a `customMenu: "custom_menu.json"` directive in your config. The file may be called whatever you want, but the name must be provided in the `config` section, and it must be stored in the Mirror's `config/` directory (same place as your `config.js` file).

An example menu is provided in this module's folder, titled `custom_menu.example.json`. You can copy this to the `/config` folder and modify as you need.

#### Key Components:

| Name | Description |
| :-: | - |
| `id` | The HTML id prefix to use for the menu item.
| `type` | The item type, either `'menu'` or `'item'`. `'menu'` is used to indicate the item is a sub-menu and has an ``items`` array. `'item'` is used for single menu items and will send a socket "REMOTE_ACTION" notification back to the server.  This requires `action:` and `content:` parameters before it can do anything.
| `text` | The text to display.  You can use the translate string `'%%TRANSLATE:YOUR_KEY_HERE%%'`, but remember to also update the appropriate file in `/translations`.
| `icon` | The [FontAwesome](https://fontawesome.com/v4.7.0/icons/) icon to use (without the leading `-fa`)
| `items` | An array of sub-menu items to use with `"type":"menu"`. Should be the same format as the top level menu (i.e. the menu structure is recursive).
| `action` | The `REMOTE_ACTION` notification action name, usually `NOTIFICATION`. Required for `"type":"item"` items to be able to do anything.
| `content` | The `REMOTE_ACTION` action payload to send.  Usually for `"NOTIFICATION"`, this is of the form `{ "notification": "NOTIFICATION_TO_SEND", "payload": "PAYLOAD_TO_SEND"}`

## License

### The MIT License (MIT)

Copyright © 2016 Joseph Bethge

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the “Software”), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

**The software is provided “as is”, without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.**


CHANGELOG.md
# MMM Remote Control Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.3.6] - 2021-08-01

### Fixed

- API now updates the modules list

## [2.3.5] - 2021-07-08

### Added

- Simplified Chinese translation
- 'PM2 not installed' warning

## [2.3.4] - 2021-04-21

### Added
- Now you can use MANAGE_CLASSES to use them between modules, instead of just the API

### Fixed
- Classes now detects when you're using identifiers and names in the same action (#259)

## [2.3.3] - 2021-04-05

### Changed
- `request` is deprecated inside MM package. Now using `node-fetch` (#257)

## [2.3.2] - 2021-02-19

### Fixed
- `value` now travels along with the payload when slide or input it's used on Custom Menus. (#251)

### Changed
- If you use slide or input, and you add a string payload, it'll now be available in the `string` object of the payload.

## [2.3.1] - 2020-12-29

### Fixed
- `Find` it's not defined inside some Electron instances (#242 and #235)
- `undefined` modules generated by the `disabled` tag are now handled. (DiscoverITLLC/MagicMirror#2382)

## [2.3.0] - 2020-12-23

### Added
- Custom Shell Commands for everyone! (#159)
- Custom Menus: User Input Field (#181) and Sliders

### Fixed
- "TV is off" now detected (#234)
- Toggle and Status Monitor working as expected (#234)

### Changed
- Now the system used for turn on and off the screen will be `vcgencmd` (#227 and more)

## [2.2.2] - 2020-11-24

### Fixed
- Module Installation now working
- iframe now working (#161)

## [2.2.1] - 2020-11-18

### Fixed
- Module Identifier now working as expected (#229)
- Update Installation seems to work

## [2.2.0] - 2020-11-16

### Fixed
- Default values now removed from backup (#12)
- Custom Menus now works as expected
- API working, not well implemented in the past
- API userPresence now working as expected

### Added
- Updates now show if there's an update available every day (#52)
- Templates for issues and PRs, and also stale for automanagement of issues.
- Close Dev Tools (#119)
- Undo Config Implementation [Beta]
- Classes to show, hide or toggle multiple modules at once (#34)
- Classes and saves API
- Changelog of every module updated
- [Showdown](https://github.com/showdownjs/showdown) implemented in order to show changelog markdown.
- secureEndpoint config to bypass the non-apikey limitation. This could be dangerous, use it with caution.
- Added POST support for Monitor API (#200)
- Added endpoint to edit config file (#206)
- Endpoint /api/docs now shows you the documentation available for the API. You can test your mirror right there!

### Changed
- **[lodash](https://lodash.com/) required**. Do `npm install` on the Remote Control module.
- Alert button don't show up when Alert module isn't active
- The way monitor turn on and off (#225)
- Now hide, show or toggle modules also accept arrays
- /api/test can be reach without any apiKey
- /api/modules/installed and /available are now /api/module/installed and /available
- ApiKey required in order to change substantial things on the Mirror
- Some Endpoints are gonna be deprecated in the future. You can see those inside /api/docs, in the Legacy menu.

### Removed
- /api/modules it's no longer available, you can use /api/module instead.
- Postman collection deprecated ~ (Sorry n.n)

## [2.1.0] - 2020-11-01

Hello! Ezequiel here. Just wanted to say thanks for trust in me, in the past days I made a lot of changes into the code, adding some functions that'll surely be in a future release, and also putting everything together in my fork. I answered almost every issue raised, and tried to help every person that use this module. Today, I'm glad to be able to share everything I learned to all of you. I apologize for some fast and uncommented commits, I just thinked that some things needed to be fixed ASAP.
See you in future commits, issues and PRs :D

### Fixed

- A typo in `es` translation
- A few typos in README.md (#134 and #149) and API/README.md (#179)
- Delayed commands should now work (#190)
- Typo on remote_action (#184)
- IP now showing (#194)
- MM restart button don't just stop anymore (#126)
- Saving config should work as expected now (#153)
- installer.sh now detects where's the node instalation (#222)

### Added

- Danish translation (#157)
- Italian translation (#162)
- Port now showing according to config.js (#98)
- Custom commands for shutdown and reboot

### Changed

- Overwrite of local changes when updating from a repository
- Now requires MagicMirror version 2.12

## [2.0.1] - 2020-10-28

**Huge thanks to [@ezeholz](https://github.com/ezeholz)** who has offered to maintain the module fron now on!
Credit for this (and future) versions and releases goes to @ezeholz (unless noted otherwise).

Now requires MagicMirror version 2.7.

### Fixed

- Path to font awesome icons
- A few typos in `ca` and `es` translations
- Updates to `remote.html` to support new `basePath` feature in MM `config.js`, [follow up to this MM issue](https://github.com/DiscoverITLLC/MagicMirror/issues/1973), related to #185

## [2.0.0] - 2019-02-21

Huge shoutout to [shbatm](https://github.com/shbatm) for his work on this new major version, which brings a new API, custom menus and commands and lots of other stuff:

### Added

- REST API interface for controlling all aspects of the MagicMirror from HTTP RESTful-style GET and POST calls, based on principles from [MMM-Api](https://github.com/juzim/MMM-Api)
    + Full API Documentation at [API/README.md](API/README.md)
- Live daily updates of `modules.json` from the MagicMirror wiki, based on principles from [MMM-Remote-Control-Repository](https://github.com/eouia/MMM-Remote-Control-Repository).  
- Incorporated some features found in [MMM-OnScreenMenu](https://github.com/shbatm/MMM-OnScreenMenu) that were not originally in this module.

installer.sh
#!/usr/bin/env bash

# This is an experimental installer script for MagicMirror2 Remote-Control

echo ""
echo "Installation for the MagicMirror2 Remote-Control module started!"
echo ""
echo "Notice: This script and the installed software is provided “as is”, without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software."
echo ""

check_yes() {
    read -p ">>> $1 [y/N]? " -n 1 REPLY
    echo ""
    if [[ ! "$REPLY" =~ ^[Yy]$ ]]; then
        return 1
    fi
    return 0
}

check_no() {
    read -p ">>> $1 [Y/n]? " -n 1 REPLY
    echo ""
    if [[ $REPLY =~ ^[Nn]$ ]]; then
        return 1
    fi
    return 0
}

if check_yes "Continue?"; then
    echo ""
else
    exit 0
fi

# assume default install location
MM_HOME=$HOME/MagicMirror
MODULE_NAME=MMM-Remote-Control
FORK=DiscoverITLLC

# check if we are correct by searching for https://github.com/DiscoverITLLC/MagicMirror in package.json
TEST_STRING="\"url\": \"git+https://github.com/DiscoverITLLC/MagicMirror.git\""
if grep -sq "$TEST_STRING" "$MM_HOME/package.json"; then
    # we found it
    echo -n ""
else
    # assume we are in the correct directory
    MM_HOME=`pwd`
    if grep -sq "$TEST_STRING" "$MM_HOME/package.json"; then
        # found it again
        echo -n ""
    else
        echo "Could not find MagicMirror2 installation directory."
        echo "Please start this script again from the MagicMirror directory."
        exit 1
    fi
fi

if [ -d "$MM_HOME/modules/$MODULE_NAME" ] ; then
    # already installed
    echo "Directory $MM_HOME/modules/$MODULE_NAME already exists."
    echo ""

    cd "$MM_HOME/modules/$MODULE_NAME"

    BRANCH="$(git symbolic-ref HEAD 2>/dev/null)" || BRANCH="(unnamed branch)" # detached HEAD

    BRANCH=${BRANCH##refs/heads/}

    echo "You are currently on the $BRANCH branch."
    echo ""
    if [ "$BRANCH" == "master" ]; then
        if check_yes "Do you want to switch to the develop branch?"; then
            git checkout develop
        fi
    else
        if check_yes "Do you want to switch to the master branch?"; then
            git checkout master
        fi
    fi
    echo ""
    if check_no "Do you want to update your branch?"; then
        echo ""
        echo "Pulling changes..."
        git pull
        if [ $? -ne 0 ]; then
            echo "Could not pull successfully."
            exit 1;
        fi
        echo ""
        echo "Checking for new dependencies to install..."
        echo ""
        npm install
        if [ $? -ne 0 ]; then
            echo "Failed to install new dependencies."
            exit 1;
        fi
        echo "Done."
        echo ""
        echo "Update finished!"
    else
        echo "Already installed, not upgrading."
    fi
else
    echo "MagicMirror2 detected in: $MM_HOME"
    echo ""
    if check_yes "Is this correct and do you want to start installation?"; then
            echo ""
            echo "You can use either the master or the develop branch."
            echo "The develop branch contains more features, but is also more likely to cause errors or crashes."
            echo "This can be changed later by executing this script again, or using the git branch command."
            echo ""
            echo "By default the master branch will be installed."
            echo ""
            if check_yes "Do you want to install the develop branch instead?"; then
                BRANCH=develop
            else
                BRANCH=master
            fi

            cd "$MM_HOME/modules"
            echo ""
            echo "Cloning the repository on $BRANCH branch..."
            echo ""
            git clone https://github.com/$FORK/$MODULE_NAME.git -b $BRANCH
            if [ $? -ne 0 ]; then
                echo "Failed. Do you have an internet connection?"
                exit 1;
            fi
            cd $MODULE_NAME

            echo ""
            echo "Installing dependencies..."
            echo ""
            npm install
            if [ $? -ne 0 ]; then
                echo "Failed to install dependencies."
                exit 1;
            fi
            echo "Done."
            echo ""
            echo "Installation finished."
    else
        echo "Installation skipped."
    fi
fi

# Get an UUID to use as an API key
NODE_BIN=$(which node)
APIKEY=$($NODE_BIN -e 'console.log(require("uuid/v4")().replace(/-/g, ""));');

echo ""
if check_no "Do you want to view instructions on how to configure the module?"; then
    echo "(1) Please add the following snippet into your modules array in your config.js:"
    echo -e "\033[33m    -------------- copy below this line --------------"
    echo -e "    {"
    echo -e "        module: '$MODULE_NAME'",
    echo -e "        // uncomment the following line to show the URL of the remote control on the mirror"
    echo -e "        // position: 'bottom_left'",
    echo -e "        // you can hide this module afterwards from the remote control itself"
    echo -e "        config: {"
    echo -e "\033[31m            apiKey: '$APIKEY'\033[33m"
    echo -e "        }"
    echo -e "    },"
    echo -e "    -------------- copy above this line --------------\033[0m"
    echo ""
    echo "(2) Also you will need to change the address at which the server listens:"
    echo "    Search for the following line in  your config.js:"
    echo "            address: \"localhost\","
    echo "    and change it to allow any device to contact your Mirror:"
    echo "            address: \"0.0.0.0\","
    echo ""
    echo "(3) Make sure to add the IPs of your devices from which you want to access the Remote Control to the ipWhitelist array."
    echo "    If for example have a phone with the IP address 192.168.0.50 search for the following line:"
    echo "            ipWhitelist: [\"127.0.0.1\", \"::ffff:127.0.0.1\", \"::1\"],"
    echo "    and add your IP to the list:"
    echo "            ipWhitelist: [\"127.0.0.1\", \"::ffff:127.0.0.1\", \"::1\", \"::ffff:192.168.0.50\"],"
    echo "    Alternatively, you can also add multiple devices with IPs such as 192.168.0.XXX:"
    echo "            ipWhitelist: [\"127.0.0.1\", \"::ffff:127.0.0.1\", \"::1\", \"::ffff:192.168.0.1/120\", \"192.168.0.1/24\"],"
    echo ""

    if check_no "Got it?"; then
        echo ""
    else
        echo ""
        echo "Please have a look at the links below for help."
        echo ""
    fi
else
    echo ""
    echo -e "\033[31mYou should also set an API key in your config section!\033[0m"
    echo "  It's dangerous to go alone! Take this. "
    echo -e "\033[31m  apiKey: '$APIKEY'\033[0m"
    echo "  I made it just for you."
    echo ""
fi
echo "Have fun with the module, if you have any problems, please search for help on github or in the forum:"
echo ""
echo "   Github : https://github.com/$FORK/$MODULE_NAME"
echo "   Forum  : http://forum.magicmirror.builders"
echo ""
echo "Do not forget to restart your MagicMirror2 to activate the module! Installation finished."
echo ""
exit 0


scripts/download_modules.py
import argparse, os, urllib2, sys, json, re

def get(url):
    try:
        connection = urllib2.urlopen(url)
        content = connection.read()
        connection.close()
        return content
    except urllib2.HTTPError, e:
        print "Error (HTTP)", e.code, url
        sys.exit(1)
    except urllib2.URLError, e:
        print "Error (URL)", e.reason, url
        sys.exit(1)

def makeConsistentName(name):
    short_name = re.sub("mmm?-", "", name, flags=re.IGNORECASE)
    short_name = short_name[0].upper() + short_name[1:]
    short_name = short_name.replace("_", " ").replace("-", " ")
    
    position = short_name.find(" ")
    while position >= 0:
        next_position = position + 1
        short_name = short_name[:next_position] + short_name[next_position].upper() + short_name[next_position + 1:]
        position = short_name.find(" ", next_position)

    short_name = re.sub("([a-z])([A-Z])", "\\1 \\2", short_name)

    return short_name

def parse(content):
    regex = re.compile("- \*\*\[(.*)\]\((.*)\)\*\*.*<br>(.*)")
    matches = re.findall(regex, content)

    module_data_list = []

    for module_author, url, desc in matches:
        name = module_author
        author = module_author
        if (" by " in module_author):
            name, author = module_author.split(" by ")

        # name needs to be normalized later again anyway
        # short_name = makeConsistentName(name)

        url = url.replace("bit.ly/MMM-Instagram", "github.com/kapsolas/MMM-Instagram")
        url = url.replace("bit.ly/MMM-Flickr", "github.com/kapsolas/MMM-Flickr")
        url = url.replace("http:", "https:")
        split_url = url.split("/")[:5]
        identifier = split_url[-2] + "/" + split_url[-1]
        url = "/".join(split_url)

        module_data = {}
        module_data["id"] = identifier
        module_data["longname"] = name
        # module_data["name"] = short_name
        module_data["author"] = author
        module_data["url"] = url
        module_data["desc"] = desc.strip()

        module_data_list.append(module_data)

    # module_data_list = sorted(module_data_list, key=lambda x: x["name"])

    print json.dumps(module_data_list, sort_keys=True, indent=4, separators=(',', ': '))

def main(args):
    content = get(args.url)
    parse(content)
    sys.exit(0)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='command line utility to download current modules')
    parser.add_argument('-u', '--url', nargs='?', default='https://raw.githubusercontent.com/wiki/DiscoverITLLC/MagicMirror/MagicMirror%C2%B2-Modules.md', help='url to extract modules from')
    parser.add_argument('-o', '--outfile', nargs='?', type=argparse.FileType('w'), default=sys.stdout, help='write output to a file (default stdout)')
    args = parser.parse_args()

    main(args)


scripts/download_modules.js
/* Magic Mirror Package List Downloader
 * Module: Remote Control
 *
 * Usage: 
 *   var downloadModules = require('./download_modules');
 *   downloadModules({ callback: function(result) { console.log(result); }});
 *      downloadModules accepts a CONFIG object which will overwrite the values
 *        in the defaults section below.
 *
 * By shbatm
 * MIT Licensed.
 */

/* jshint esversion:6, node: true */
const fetch = require("node-fetch");
const path = require("path");
const fs = require("fs");
const util = require("util");

var downloadModules = {
    defaults: {
        modulesFile: path.resolve(__dirname, "../modules.json"),    // Path to modules file
        sourceUrl: 'https://raw.githubusercontent.com/wiki/DiscoverITLLC/MagicMirror/3rd-Party-Modules.md', // Source url
        refreshRate: 24 * 3600,                                     // Max Refresh of One Day
        force: false,                                               // Force the update
        callback: function(result) { console.log(result); }         // Callback to run on success or failure
    },

    init: function(config) {
        if (!config) { config = {}; }
        this.config = Object.assign({}, this.defaults, config);

        return this;
    },

    parseList: function(content) {
        let re = /\|\s?\[(.*?)\]\((.*?)\)\s?\|(.*?)\|(.*)\|?/g;
        let modules = [];

        content.match(re).forEach((line) => {
            line.replace(re, (match, name, url, author, desc) => {
                let modDetail = {
                    longname: name.trim(),
                    id: url.replace(".git", "").replace(/.*\/(.*?\/.*?)$/, "$1").trim(),
                    url: url.replace(".git", "").trim(),
                    author: author.replace(/\[(.*)\]\(.*\)/, "$1").trim(),
                    desc: desc.replace(/\|/, "").trim()
                };
                modules.push(modDetail);
            });
        });

        return modules;
    },

    getPackages: function() {
        fetch(this.config.sourceUrl)
        .then(response => {
            if (response.status === 200) {
                return response;
            } else if (response.statusCode === 401) {
                console.error("MODULE LIST ERROR: Could not load module data from wiki. 401 Error");
                this.config.callback("ERROR_401");
                return;
            } else {
                console.error("MODULE LIST ERROR: Could not load data.", statusText);
                this.config.callback("ERROR_LOADING_DATA");
                return;
            }
        })
        .then(response => response.text())
        .then(body => {
            let modules = this.parseList(body);
            var json = JSON.stringify(modules, undefined, 2);
            var jsonPath = this.config.modulesFile;
            fs.writeFile(jsonPath, json, "utf8", (err, data) => {
                if (err) {
                    console.error("MODULE LIST ERROR: modules.json updating fail:" + err.message);
                    this.config.callback("ERROR_UPDATING");
                } else {
                    this.config.callback("UPDATED");
                }
            });
        })
        .catch(error => {
            console.error("MODULE LIST ERROR: Could not load data.", error);
            this.config.callback("ERROR_LOADING_DATA");
            return;
        });
        return;
    },

    checkLastModified: function() {
        fs.stat(this.config.modulesFile, (err, stats) => {
            let mtime = Math.round(new Date(util.inspect(stats.mtime)).getTime() / 1000);
            let updatedAfter = new Date(Math.round(new Date().getTime() / 1000) - this.config.refreshRate).getTime();
            let needsUpdate = mtime <= updatedAfter;
            if (needsUpdate || this.config.force) { 
                this.getPackages(); 
            } else {
                this.config.callback("NO_UPDATE_REQUIRED");
            }
        });
        return;
    }
};

if (typeof module !== "undefined") {
    module.exports = function(config) {
        downloadModules.init(config);
        downloadModules.checkLastModified();
        return;
    };
}